src/Distortion.cpp

```diff
@@ -162,93 +83,144 @@ std::tuple<double, double> distortionFunction(double ux, double uy,
- * focal plane coordinate. This method works by iteratively adding distortion
- * until the new distorted point, r, undistorts to within a tolerance of the
- * original point, rp.
- *
- * @param inFocalPlaneX Distorted x
- * @param inFocalPlaneY Distorted y
- * @param opticalDistCoef Distortion coefficients
- * @param desiredPrecision Convergence precision
- * @param tolerance Tolerance of r^2
- *
- * @returns undistortedPoint Newly adjusted focal plane coordinates as an x, y tuple
- */
-std::tuple<double, double> invertDistortion(double inFocalPlaneX, double inFocalPlaneY,
-  const double opticalDistCoef[3], double desiredPrecision, double tolerance) {
-  double rp2 = (inFocalPlaneX * inFocalPlaneX) +
-               (inFocalPlaneY * inFocalPlaneY);
-  std::tuple<double, double> undistortedPoint;
-
-  if (rp2 > tolerance) {
-    double rp = sqrt(rp2);
-    // Compute first fractional distortion using rp
-    double drOverR = opticalDistCoef[0]
-                  + (rp2 * (opticalDistCoef[1] + (rp2 * opticalDistCoef[2])));
-    // Compute first distorted point estimate, r
-    double r = rp + (drOverR * rp);
-    double r_prev, r2_prev;
-    int iteration = 0;
-    do {
-      // Don't get in an end-less loop.  This algorithm should
-      // converge quickly.  If not then we are probably way outside
-      // of the focal plane.  Just set the distorted position to the
-      // undistorted position. Also, make sure the focal plane is less
-      // than 1km, it is unreasonable for it to grow larger than that.
-      if (iteration >= 15 || r > 1E9) {
-        drOverR = 0.0;
-        break;
+void removeDistortion(double dx, double dy, double &ux, double &uy,
+                      const std::vector<double> opticalDistCoeffs,
+                      DistortionType distortionType,
+                      double desiredPrecision) {
+  ux = dx;
+  uy = dy;
+
+  switch (distortionType) {
+    // Compute undistorted focal plane coordinate given a distorted
+    // coordinate set and the distortion coefficients
+    case RADIAL: {
+      double rr = dx * dx + dy * dy;
+      double tolerance = 1.0E-6;
+
+      if (rr > tolerance)
+      {
+        double dr = opticalDistCoeffs[0] + (rr * (opticalDistCoeffs[1] + rr * opticalDistCoeffs[2]));
+        ux = dx * (1.0 - dr);
+        uy = dy * (1.0 - dr);
+      }
+    }
+    break;
+    // Computes undistorted focal plane (x,y) coordinates given a distorted focal plane (x,y)
+    // coordinate. The undistorted coordinates are solved for using the Newton-Raphson
+    // method for root-finding if the distortionFunction method is invoked.
+    case TRANSVERSE: {
+      // Solve the distortion equation using the Newton-Raphson method.
+      // Set the error tolerance to about one millionth of a NAC pixel.
+      const double tol = 1.4E-5;
+
+      // The maximum number of iterations of the Newton-Raphson method.
+      const int maxTries = 60;
```

Comment:
Same question here about the constant.

Reply:
Hard coding a max iterations should be fine. 60 is kind of large number of max iterations. Newton-Raphson should have quadratic convergence. A small number of iterations will get us incredibly close. Something like 20 is probably better.